driver: watchdog: Add provision to add early warning interrupt

ghostMan-pac opened the pull request:
- Added new structure element to keep early warning interrupt callback for watchdog window open.
- This is added so that user can initialise a callback in the interrupt ISR do some handling when the early warning window is enabled, if the hardware supports it.
- If a certain hw does not support it, they can keep it as NULL.
- A new kconfig property added to conditionally compile if the feature is available for a particular device.

ghostMan-pac committed:
driver: watchdog: Add provision to add early warning interrupt

- Added new structure element to keep early warning interrupt callback
  for watchdog window open.
- This is added so that user can initialise a callback in the
  interrupt ISR do some handling when the early warning window is
  enabled, if the hardware supports it.
- If a certain hw does not support it, they can keep it as NULL.
- A new kconfig property added to conditionally compile if the feature
  is available for a particular device.
- Made the newly added Kconfig symbol to be visible

Signed-off-by: Muhammed Asif <asifp3104@gmail.com>

pdgendt commented on include/zephyr/drivers/watchdog.h:
```diff
@@ -86,12 +86,27 @@ struct wdt_window {
+ * @param dev Watchdog device instance.
+ * @param channel_id Channel identifier.
+ */
+typedef void (*wdt_ew_callback_t)(const struct device *dev, int channel_id);
```
The type is the same as `wdt_callback_t`, no need to have a new `typedef` IMO.

ghostMan-pac replied:
you are right. they are the same type. i have used `wdt_callback_t` itself for the new variable

ghostMan-pac commented on drivers/watchdog/Kconfig:
```diff
@@ -65,6 +65,16 @@ config WDT_COUNTER_CH_COUNT
+	  supports early warning interrupts, a callback can be initialised from
+	  the user to handle the cases for when watchdog window is opened. Certain
+	  devices have the provision for an interrupt when the minimum window time is
+	  reached, in order to notify the user that the window is opened.
```
i have updated this

pdgendt commented on drivers/watchdog/Kconfig:
```diff
@@ -65,6 +65,14 @@ config WDT_COUNTER_CH_COUNT
+config WDT_EARLY_WARNING_INTERRUPT_SUPPORT
+	bool
+	help
+	  Add support for the watchdog early warning interrupt. If the watchdog peripheral
+	  supports early warning interrupts, a callback can be set by the user to handle the cases when the watchdog window is opened. Certain
+	  devices have to be provisioned for when the minimum window time is
+	  reached, in order to notify the user that the window is opened.
```
Try to limit the line length to max 100 chars.

ghostMan-pac replied:
updated

pdgendt commented on include/zephyr/drivers/watchdog.h:
```diff
@@ -92,6 +92,11 @@ struct wdt_timeout_cfg {
 	wdt_callback_t callback;
+#if defined(CONFIG_WDT_EARLY_WARNING_INTERRUPT_SUPPORT) || defined(__DOXYGEN__)
+	/** Early warning callback (can be `NULL`). */
+	wdt_callback_t ew_callback;
```
Did you delete code, as this no longer appears to be used?

ghostMan-pac replied:
I have only deleted the part where the typedef was done, as you suggested.

martinjaeger commented:
@ghostMan-pac Do you have any example hardware which supports this feature? Normally, we don't add APIs which are not used by any in-tree driver.

ghostMan-pac commented:
> @ghostMan-pac Do you have any example hardware which supports this feature? Normally, we don't add APIs which are not used by any in-tree driver.

Yes. The atmel sam0 actually have this feature. (same54_xpro board). Shall I add the changes for the sam0 watchdog driver and add it with this pr? or should I open a new one?

pdgendt commented:
> Yes. The atmel sam0 actually have this feature. (same54_xpro board). Shall I add the changes for the sam0 watchdog driver and add it with this pr? or should I open a new one?

Add it to this PR in a separate commit, please.

ghostMan-pac commented:
> > Yes. The atmel sam0 actually have this feature. (same54_xpro board). Shall I add the changes for the sam0 watchdog driver and add it with this pr? or should I open a new one?
> 
> Add it to this PR in a separate commit, please.

Sure. I will add that and update soon.

ghostMan-pac committed:
driver : watchdog : Updated Sam0 to use early warning callback

- Modified sam0 driver to use the early warning interrupt callback
instead of normal timeout callback, as SAME54 has only support for
early warning interrupt

Signed-off-by: Muhammed Asif <asifp3104@gmail.com>

ghostMan-pac committed:
samples: Watchdog: Added support for early warning callback

- Added the changes for initialising a user early warning callback

Signed-off-by: Muhammed Asif <asifp3104@gmail.com>

pdgendt commented on drivers/watchdog/wdt_sam0.c:
```diff
@@ -90,8 +90,8 @@ static void wdt_sam0_isr(const struct device *dev)
 
-	if (data->cb != NULL) {
-		data->cb(dev, 0);
+	if (data->ew_cb != NULL) {
+		data->ew_cb(dev, 0);
```
Why change the existing watchdog callback? Now this driver doesn't do anything, unless `CONFIG_WDT_EARLY_WARNING_INTERRUPT_SUPPORT` is selected. I would hope a driver always uses the existing callback function, and the "early" one would only be an additional one.

ghostMan-pac replied:
That's right. Driver will use early warning interrupt only if it is selected. The thing with early warning interrupt is that only a couple of peripherals are supporting this interrupt callback, that is why i made it `no` as default in the `kconfig` file.

The existing watchdog callback is changed because in the same54 board, the soc only supports early warning interrupt in the watchdog peripheral. 

As I understand the timeout callback is intended to get invoked only if the `WDT_FLAG_RESET_NONE` is provided to watchdog driver and the timer value is not cleared before the set timeout.

If the flag is not given to the watchdog driver, board will be reset before that callback is invoked. This flag also is not supported in same54.

nandojve commented on include/zephyr/drivers/watchdog.h:
```diff
@@ -92,6 +92,11 @@ struct wdt_timeout_cfg {
 	wdt_callback_t callback;
+#if defined(CONFIG_WDT_EARLY_WARNING_INTERRUPT_SUPPORT) || defined(__DOXYGEN__)
+	/** Early warning callback (can be `NULL`). */
+	wdt_callback_t ew_callback;
```
Do you know if there is any hardware that support both callback to be active?
	wdt_callback_t callback;
	wdt_callback_t ew_callback;

Otherwise we can still keep only one, right?

nandojve commented on drivers/watchdog/Kconfig:
```diff
@@ -65,6 +65,15 @@ config WDT_COUNTER_CH_COUNT
 
 endif # WDT_COUNTER
 
+config WDT_EARLY_WARNING_INTERRUPT_SUPPORT
```
This at end is a HW config, right? So, could it be defined in devicetree as a bool value ?

nandojve commented on samples/drivers/watchdog/src/main.c:
```diff
@@ -83,6 +83,23 @@ static void wdt_callback(const struct device *wdt_dev, int channel_id)
 
+#if WDT_EARLY_WARNING_INTERRUPT_SUPPORT
+
+static void wdt_ew_callback(const struct device *wdt_dev, int channel_id)
```
if the devictree bool option is valid, we can select the correct callback inspecting that devicetree option.
